crates/datasources/src/cassandra/mod.rs

```diff
@@ -0,0 +1,167 @@
+mod builder;
+mod errors;
+mod exec;
+use async_stream::stream;
+use async_trait::async_trait;
+use datafusion::arrow::array::{
+    ArrayRef, Float32Builder, Float64Builder, Int16Builder, Int32Builder, Int8Builder,
+    StringBuilder, TimestampMillisecondBuilder,
+};
+use datafusion::arrow::datatypes::{
+    DataType, Field, Fields, Schema as ArrowSchema, SchemaRef as ArrowSchemaRef, TimeUnit,
+};
+
+use datafusion::arrow::record_batch::RecordBatch;
+use datafusion::datasource::TableProvider;
+use datafusion::error::{DataFusionError, Result as DatafusionResult};
+use datafusion::execution::context::SessionState;
+use datafusion::execution::TaskContext;
+use datafusion::logical_expr::{Expr, TableProviderFilterPushDown, TableType};
+use datafusion::physical_expr::PhysicalSortExpr;
+use datafusion::physical_plan::metrics::{ExecutionPlanMetricsSet, MetricsSet};
+use datafusion::physical_plan::{
+    DisplayAs, DisplayFormatType, ExecutionPlan, Partitioning, RecordBatchStream,
+    SendableRecordBatchStream, Statistics,
+};
+use datafusion_ext::metrics::DataSourceMetricsStreamAdapter;
+pub use errors::*;
+use futures::Stream;
+use futures::StreamExt;
+use scylla::frame::response::result::{ColumnType, CqlValue, Row};
+
+use scylla::transport::session::Session;
+use scylla::SessionBuilder;
+use std::any::Any;
+
+use std::fmt;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use self::exec::CassandraExec;
+
+struct CassandraAccess {
+    session: Session,
+}
+
+fn try_convert_dtype(ty: &ColumnType) -> Result<DataType> {
+    Ok(match ty {
+        ColumnType::Ascii => DataType::Utf8,
+        ColumnType::Date => DataType::Date64,
```

Comment:
They use a u32 to hold the date. It's not a 1:1 with a date32. so we have to do some logic to convert it to datetime first.

Reply:
https://docs.rs/datafusion/latest/datafusion/common/arrow/datatypes/struct.Date32Type.html and https://cassandra.apache.org/doc/stable/cassandra/cql/types.html#dates / https://opensource.docs.scylladb.com/stable/cql/types.html#working-with-dates disagree

Reply:
CQL
> with “the epoch” at the center of the range (2^31)

Datafusion
> A 32-bit date type representing the elapsed time since UNIX epoch in days(32 bits).

they are not a 1:1 mapping.

Reply:
> CQL
> 
> > with “the epoch” at the center of the range (2^31)
> 
> Datafusion
> 
> > A 32-bit date type representing the elapsed time since UNIX epoch in days(32 bits).
> 
> they are not a 1:1 mapping. I thought they were at first too, but they are not. A conversion needs to happen.
> 
> The logic in place works & there are tests to confirm.

Let's leave a comment on this in the code to avoid confusion here.